udp-echo/src/main.rs

```diff
@@ -1,42 +1,58 @@
-use actix::{Actor, AsyncContext, Context, Message, StreamHandler};
+use actix::io::SinkWrite;
+use actix::{Actor, Arbiter, AsyncContext, Context, Message, StreamHandler};
+use bytes::Bytes;
 use bytes::BytesMut;
 use futures::stream::SplitSink;
-use futures::{Future, Sink, Stream};
+use futures_util::stream::StreamExt;
+use std::io::Result;
 use std::net::SocketAddr;
-use tokio::codec::BytesCodec;
-use tokio::net::{UdpFramed, UdpSocket};
+use tokio::net::UdpSocket;
+use tokio_util::codec::BytesCodec;
+use tokio_util::udp::UdpFramed;
+
+type SinkItem = (Bytes, SocketAddr);
+type UdpSink = SplitSink<UdpFramed<BytesCodec>, SinkItem>;
 
 struct UdpActor {
-    sink: SplitSink<UdpFramed<BytesCodec>>,
+    sink: SinkWrite<SinkItem, UdpSink>,
 }
 impl Actor for UdpActor {
     type Context = Context<Self>;
 }
 
 #[derive(Message)]
+#[rtype(result = "()")]
 struct UdpPacket(BytesMut, SocketAddr);
-impl StreamHandler<UdpPacket, std::io::Error> for UdpActor {
+
+impl StreamHandler<UdpPacket> for UdpActor {
     fn handle(&mut self, msg: UdpPacket, _: &mut Context<Self>) {
         println!("Received: ({:?}, {:?})", msg.0, msg.1);
-        (&mut self.sink).send((msg.0.into(), msg.1)).wait().unwrap();
+
+        self.sink.write(("PING\n".into(), msg.1)).unwrap();
     }
 }
 
+impl actix::io::WriteHandler<std::io::Error> for UdpActor {}
+
 fn main() {
     let sys = actix::System::new("echo-udp");
 
-    let addr: SocketAddr = "127.0.0.1:0".parse().unwrap();
-    let sock = UdpSocket::bind(&addr).unwrap();
-    println!(
-        "Started udp server on: 127.0.0.1:{:?}",
-        sock.local_addr().unwrap().port()
-    );
-
-    let (sink, stream) = UdpFramed::new(sock, BytesCodec::new()).split();
-    UdpActor::create(|ctx| {
-        ctx.add_stream(stream.map(|(data, sender)| UdpPacket(data, sender)));
-        UdpActor { sink: sink }
+    Arbiter::spawn(async {
```

Comment:
How about adding `actix-rt` as a dependency and using `actix_rt::main` for an `async` main instead of spawning it as a background future? I think it would be a clearer code and more in tune to how actix apps usually looks like on newer releases.

Reply:
I tried that, but there is not really a `future` to `await` for. So it shuts down immediately after it starts up.

I changed it to use `actix-rt` and at the end of main it awaits on `actix_rt::Arbiter::local_join()` but I don't think it is a nice solution. Any better idea?

Reply:
Being able to make the runtime `await` for internally spawned futures, the `UdpActor` actor in this case, is the intention of the `Arbiter::local_join`. I'm fine with it.